rust/connlib/tunnel/src/client.rs

```diff
@@ -1286,315 +1297,245 @@ mod tests {
         );
-        assert_eq!(hashset(client_state.routes()), expected_routes(vec![]));
     }
 
-    #[test]
-    fn set_resource_works() {
+    #[test_strategy::proptest]
+    fn adding_same_resource_with_different_address_updates_the_address(
+        #[strategy(cidr_resource())] resource: ResourceDescriptionCidr,
+        #[strategy(ip_network())] new_address: IpNetwork,
+    ) {
         let mut client_state = ClientState::for_test();
+        client_state.add_resources(&[ResourceDescription::Cidr(resource.clone())]);
 
-        client_state.set_resources(vec![
-            cidr_foo_resource("10.0.0.0/24"),
-            dns_bar_resource("baz.com"),
-        ]);
+        let updated_resource = ResourceDescriptionCidr {
+            address: new_address,
+            ..resource
+        };
+
+        client_state.add_resources(&[ResourceDescription::Cidr(updated_resource.clone())]);
 
         assert_eq!(
             hashset(client_state.resources().iter()),
-            hashset(
-                [
-                    cidr_foo_resource("10.0.0.0/24"),
-                    dns_bar_resource("baz.com")
-                ]
-                .iter()
-            )
+            hashset(&[ResourceDescription::Cidr(updated_resource),])
         );
         assert_eq!(
-            HashSet::<IpNetwork>::from_iter(client_state.routes()),
-            expected_routes(vec![IpNetwork::from_str("10.0.0.0/24").unwrap()])
+            hashset(client_state.routes()),
+            expected_routes(vec![new_address])
         );
     }
 
-    #[test]
-    fn set_resource_replaces_old_resources() {
+    #[test_strategy::proptest]
+    fn adding_cidr_resource_with_same_id_as_dns_resource_replaces_dns_resource(
+        #[strategy(dns_resource())] resource: ResourceDescriptionDns,
+        #[strategy(ip_network())] address: IpNetwork,
+    ) {
         let mut client_state = ClientState::for_test();
+        client_state.add_resources(&[ResourceDescription::Dns(resource.clone())]);
 
-        client_state.set_resources(vec![
-            cidr_foo_resource("10.0.0.0/24"),
-            dns_bar_resource("baz.com"),
-        ]);
-        client_state.set_resources(vec![cidr_baz_resource("11.0.0.0/24")]);
+        let dns_as_cidr_resource = ResourceDescriptionCidr {
+            address,
+            id: resource.id,
+            name: resource.name,
+        };
+
+        client_state.add_resources(&[ResourceDescription::Cidr(dns_as_cidr_resource.clone())]);
 
         assert_eq!(
             hashset(client_state.resources().iter()),
-            hashset([cidr_baz_resource("11.0.0.0/24")].iter())
+            hashset(&[ResourceDescription::Cidr(dns_as_cidr_resource),])
         );
         assert_eq!(
             hashset(client_state.routes()),
-            expected_routes(vec![IpNetwork::from_str("11.0.0.0/24").unwrap()])
+            expected_routes(vec![address])
         );
     }
 
-    #[test]
-    fn set_resource_updates_old_resource_with_same_id() {
```

Comment:
@conectado Are you okay with this and the following test being removed? Whilst I think it is important to test the properties of `set_resources`, these tests don't actually do that because they would also pass if we would change the implementation to "clear all and add new resources". Thus I think they are redundant.

Reply:
While it doesn't test that resources are "stable" this does test that even in the case that a resource is set with the same id it does get replaced. 

With the current implementation this is a mistake that can happen easily, since we remove/add based on ids I think it's good to keep the test.

Might be worth adding a clarifying comment.

Reply:
Also thanks for pointing this out! I did miss this

Reply:
> While it doesn't test that resources are "stable" this does test that even in the case that a resource is set with the same id it does get replaced.
> 
> With the current implementation this is a mistake that can happen easily, since we remove/add based on ids I think it's good to keep the test.

But how can I assert that it gets replaced?

I have a test that asserts that after I use `set_resources`, we have the new set of resources (and not the old ones). How is a test that starts with the same resource different from that?

Reply:
> Also thanks for pointing this out! I did miss this

Did you review commit-by-commit? I tried to make it very obvious in the comments!

Reply:
this one I did it as a whole thing

Reply:
> I have a test that asserts that after I use `set_resources`, we have the new set of resources (and not the old ones). How is a test that starts with the same resource different from that?

@conectado ^

Reply:
> I have a test that asserts that after I use set_resources, we have the new set of resources (and not the old ones). How is a test that starts with the same resource different from that?

By making sure that there was a resource before with the same ID but different CIDR then we can assert that the routes have been replaced to the ones in the new resource, this way we make partially sure that the old resource was removed and the new one added.

Reply:
Can you write the test? I don't think I get how it is different from the one we have.

Reply:
the difference from the one we already have is that `cidr_resource1` and `cidr_resource2` needs the same id

give me 1 minute I'll push a new test quickly to discuss there

Reply:
here it is https://github.com/firezone/firezone/pull/4503/commits/806456f8d016d061b1d172f47099f944dfd62e55